node/src/main/scala/com/wavesplatform/transaction/TxValidationError.scala

```diff
@@ -47,94 +48,87 @@ object TxValidationError {
+      FailedTransactionError(Cause.AssetScriptInAction, spentComplexity, log, Some(error), Some(assetId))
 
-    private val target: String = this match {
-      case _: ByAssetScript         => "Asset"
-      case _: ByAssetScriptInAction => "Asset"
-      case _: ByAccountScript       => "Account"
-      case _: FailedByDAppScript    => "DApp"
-      case _: RejectedByDAppScript  => "DApp"
-    }
+    def notAllowedByAssetInAction(spentComplexity: Long, log: Log[Id], assetId: ByteStr): FailedTransactionError =
+      FailedTransactionError(Cause.AssetScriptInAction, spentComplexity, log, None, Some(assetId))
 
-    override def toString: String = s"ScriptExecutionError(error = $error, type = $target, log =${logToString(log)})"
-  }
+    def assetExecution(error: String, spentComplexity: Long, log: Log[Id], assetId: ByteStr): FailedTransactionError =
+      FailedTransactionError(Cause.AssetScript, spentComplexity, log, Some(error), Some(assetId))
 
-  object ScriptExecutionError {
-    def apply(error: String, log: Log[Id], assetId: Option[ByteStr]): ScriptExecutionError =
-      assetId.fold[ScriptExecutionError](ByAccountScript(error, log))(ai => ByAssetScript(error, log, ai))
+    def notAllowedByAsset(spentComplexity: Long, log: Log[Id], assetId: ByteStr): FailedTransactionError =
+      FailedTransactionError(Cause.AssetScript, spentComplexity, log, None, Some(assetId))
 
-    final case class ByAccountScript(error: String, log: Log[Id])                         extends ScriptExecutionError
-    final case class RejectedByDAppScript(error: String, log: Log[Id] = List.empty)       extends ScriptExecutionError
-    final case class FailedByDAppScript(error: String, log: Log[Id] = List.empty)         extends FailedExecutionError
-    final case class ByAssetScriptInAction(error: String, log: Log[Id], assetId: ByteStr) extends FailedExecutionError
-    final case class ByAssetScript(error: String, log: Log[Id], assetId: ByteStr)         extends FailedExecutionError
-  }
-
-  case class InsufficientInvokeActionFee(error: String) extends ValidationError with FailedTransactionError {
-    override def toString: String = s"InsufficientInvokeActionFee(error = $error)"
-  }
-
-  sealed trait TransactionNotAllowedByScript extends ValidationError {
-    import TransactionNotAllowedByScript._
-
-    def log: Log[Id]
-    val isAssetScript: Boolean = this match {
-      case _: ByAssetScript         => true
-      case _: ByAssetScriptInAction => true
-      case _: ByAccountScript       => false
+    sealed trait Cause extends Product with Serializable {
+      def code: Int
     }
-
-    private val target: String = this match {
-      case _: ByAssetScript         => "Asset"
-      case _: ByAssetScriptInAction => "Asset"
-      case _: ByAccountScript       => "Account"
+    object Cause {
+      case object DAppExecution extends Cause {
```

Comment:
what is the difference with AssetScriptInAction and AssetScript?
And DAppExecution and FeeForActions
may be rename error codes?

Reply:
AssetScriptInAction is for asset scripts executed in dApp actions, AssetScript is for remaining asset scripts.
DAppExecution is for errors throwed by dApp script, FeeForActions is for case when there is not enough fee for all actions in dApp.

I added .toString representation of FailedTransactionError: https://github.com/wavesplatform/Waves/blob/125626afcd91216ad5969719673fee0969966ba3/node/src/main/scala/com/wavesplatform/transaction/TxValidationError.scala#L79